plugins/discourse-presence/plugin.rb

```diff
@@ -22,154 +22,50 @@ class Engine < ::Rails::Engine
-    end
-
-    # return true if a key was added
-    def self.add(type, id, user_id)
-      key = get_redis_key(type, id)
-      result = Discourse.redis.hset(key, user_id, Time.zone.now)
-      Discourse.redis.expire(key, MAX_BACKLOG_AGE)
-      result
-    end
-
-    # return true if a key was deleted
-    def self.remove(type, id, user_id)
-      key = get_redis_key(type, id)
-      Discourse.redis.expire(key, MAX_BACKLOG_AGE)
-      Discourse.redis.hdel(key, user_id) > 0
-    end
-
-    def self.get_users(type, id)
-      user_ids = Discourse.redis.hkeys(get_redis_key(type, id)).map(&:to_i)
-      User.where(id: user_ids)
-    end
-
-    def self.publish(type, id)
-      users = get_users(type, id)
-      serialized_users = users.map { |u| BasicUserSerializer.new(u, root: false) }
-      message = { users: serialized_users, time: Time.now.to_i }
-      messagebus_channel = get_messagebus_channel(type, id)
-
-      topic = type == 'post' ? Post.find_by(id: id).topic : Topic.find_by(id: id)
-
-      if topic.private_message?
-        user_ids = User.where('admin OR moderator').pluck(:id) + topic.allowed_users.pluck(:id)
-        group_ids = topic.allowed_groups.pluck(:id)
-
-        MessageBus.publish(
-          messagebus_channel,
-          message.as_json,
-          user_ids: user_ids,
-          group_ids: group_ids,
-          max_backlog_age: MAX_BACKLOG_AGE
-        )
-      else
-        MessageBus.publish(
-          messagebus_channel,
-          message.as_json,
-          group_ids: topic.secure_group_ids,
-          max_backlog_age: MAX_BACKLOG_AGE
-        )
-      end
-
-      users
-    end
-
-    def self.cleanup(type, id)
-      has_changed = false
-
-      # Delete entries older than 20 seconds
-      hash = Discourse.redis.hgetall(get_redis_key(type, id))
-      hash.each do |user_id, time|
-        if Time.zone.now - Time.parse(time) >= 20
-          has_changed |= remove(type, id, user_id)
-        end
-      end
-
-      has_changed
-    end
-
-  end
-
   require_dependency "application_controller"
 
   class Presence::PresencesController < ::ApplicationController
     requires_plugin PLUGIN_NAME
     before_action :ensure_logged_in
-
-    ACTIONS ||= [-"edit", -"reply"].freeze
-
-    def publish
-      raise Discourse::NotFound if current_user.blank? || current_user.user_option.hide_profile_and_presence?
-
-      data = params.permit(
-        :response_needed,
-        current: [:action, :topic_id, :post_id],
-        previous: [:action, :topic_id, :post_id]
+    before_action :ensure_presence_enabled
+
+    def handle_message
+      topic_id = permitted_params[:topic_id]
+      topic = Topic.find_by(id: topic_id)
+      guardian.ensure_can_see!(topic)
+
+      MessageBus.publish(
+        "/presence/#{topic_id}",
+        {
+          user: BasicUserSerializer.new(current_user, root: false).as_json,
+          state: permitted_params[:state]
+        },
+        group_ids: topic.secure_group_ids,
```

Comment:
if it is a whisper you are editing group_ids is only going to be the intersection of "staff" with topic secure groups.

Also the blew away all support for pms... you are going to need to reconstruct that ... topic allowed users / groups for messages.

Reply:
Yup I've not tackled that yet. I've only gotten replying and closing of the composer to publish the right messages so far.

Reply:
no worries at all, the approach here looks a-ok to me.